src/main/scala/com/microsoft/hyperspace/actions/CreateActionBase.scala

```diff
@@ -163,4 +161,44 @@ private[actions] abstract class CreateActionBase(dataManager: IndexDataManager)
             s"from available source columns '${dfColumnNames.mkString(",")}'")
     }
   }
+
+  private def getIndexDataFrame(
+      spark: SparkSession,
+      df: DataFrame,
+      indexConfig: IndexConfig): (DataFrame, Seq[String], Seq[String]) = {
+    val (resolvedIndexedColumns, resolvedIncludedColumns) = resolveConfig(df, indexConfig)
+    val indexConfigColumns = resolvedIndexedColumns ++ resolvedIncludedColumns
+    val addLineage = spark.sessionState.conf
+      .getConfString(
+        IndexConstants.INDEX_LINEAGE_ENABLED,
+        IndexConstants.INDEX_LINEAGE_ENABLED_DEFAULT)
+      .toBoolean
+
+    val indexDF = if (addLineage) {
+      // Lineage is captured using two sets of columns:
+      // 1. DATA_FILE_NAME_COLUMN column contains source data filename for each index record.
+      // 2. If source data is partitioned, all partitioning key(s) are added to index schema.
+      val additionalPartitioningColumns = getPartitioningColumns(df).filter(
+        ResolverUtils.resolve(spark, _, indexConfigColumns).isEmpty)
+      val dataFrameColumns = indexConfigColumns ++ additionalPartitioningColumns
+      df.select(dataFrameColumns.head, dataFrameColumns.tail: _*)
+        .withColumn(IndexConstants.DATA_FILE_NAME_COLUMN, getFileName(input_file_name()))
+    } else {
+      df.select(indexConfigColumns.head, indexConfigColumns.tail: _*)
+    }
+
+    (indexDF, resolvedIndexedColumns, resolvedIncludedColumns)
+  }
+
+  private def getPartitioningColumns(df: DataFrame): Seq[String] = {
+    // Extract partitioning keys, if original data is partitioned.
+    val partitionSchema = df.queryExecution.optimizedPlan.collect {
+      case LogicalRelation(HadoopFsRelation(_, pSchema, _, _, _, _), _, _, _) => pSchema
+    }
+
```

Comment:
Can we `assert(partitionSchemas == 1)`? (Note that it should be `val partitionSchemas` since `collect` returns `Seq`.

Reply:
Done.

Reply:
Done.